# Global fertility rate data analysis

## Table of Contents
<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#wrangling">Data Wrangling</a></li>
<li><a href="#eda">Exploratory Data Analysis</a></li>
<li><a href="#conclusions">Conclusions</a></li>
</ul>

## Introduction <a id='intro'></a>

This project aims at exploring global fertility data that was collected from the [Gapminder project](www.gapminder.org). Apart from the fertility data itself, I downloaded four other datasaets with indicators that I assume might be connected to the fertility: 

* child mortality (0-5 year olds dying per 1000 born)
* income per person (GDP/capita, PPP\$ inflation-adjusted)
* women's education (average years spent in school by women aged between 15 and 44 years)
* urban population (\% of total population living in urban areas)

In order to categorise the countries contained in the data as regions, I also downloaded [geographical data](https://www.gapminder.org/data/geo/) including these categories.

The idea is to explore how the fertility rate changed over the past 220 years in the different regions of the world. Furthermore I would like to find out, if the above indicators are correlated with the fertility rate and if the correlation is different depending on the country and/or region investigated. 

In [26]:
# import packages
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

<a id='wrangling'></a>
## Data Wrangling

### General Properties

In [86]:
# load fertility data
df_fertility = pd.read_csv('fertility.csv')
df_fertility.head()

country  1800  1801  1802  1803  1804  1805  1806  1807  1808  \
0          Afghanistan  7.00  7.00  7.00  7.00  7.00  7.00  7.00  7.00  7.00   
1              Albania  4.60  4.60  4.60  4.60  4.60  4.60  4.60  4.60  4.60   
2              Algeria  6.99  6.99  6.99  6.99  6.99  6.99  6.99  6.99  6.99   
3               Angola  6.93  6.93  6.93  6.93  6.93  6.93  6.93  6.94  6.94   
4  Antigua and Barbuda  5.00  5.00  4.99  4.99  4.99  4.98  4.98  4.97  4.97   

   ...  2091  2092  2093  2094  2095  2096  2097  2098  2099  2100  
0  ...  1.74  1.74  1.74  1.74  1.74  1.74  1.74  1.74  1.74  1.74  
1  ...  1.78  1.78  1.78  1.79  1.79  1.79  1.79  1.79  1.79  1.79  
2  ...  1.86  1.86  1.86  1.86  1.86  1.86  1.86  1.86  1.86  1.86  
3  ...  2.54  2.52  2.50  2.48  2.47  2.45  2.43  2.42  2.40  2.40  
4  ...  1.81  1.81  1.81  1.81  1.81  1.81  1.81  1.82  1.82  1.82  

[5 rows x 302 columns]

In [28]:
# check data types
df_fertility.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 184 entries, 0 to 183
Columns: 302 entries, country to 2100
dtypes: float64(301), object(1)
memory usage: 434.2+ KB


The above output shows that the countries are stored as strings and the fertility rates are stored as floats for all countries. No adjustments needed. 
***

In [29]:
# check for missing values
sum(df_fertility.isnull().any())

0

The above output shows that there is no missing values in the data.
***

In [87]:
# load child mortality data
df_mortality = pd.read_csv('child_mortality.csv')
df_mortality.head()

country   1800   1801   1802   1803   1804   1805   1806   1807   1808  \
0  Afghanistan  469.0  469.0  469.0  469.0  469.0  469.0  470.0  470.0  470.0   
1      Albania  375.0  375.0  375.0  375.0  375.0  375.0  375.0  375.0  375.0   
2      Algeria  460.0  460.0  460.0  460.0  460.0  460.0  460.0  460.0  460.0   
3      Andorra    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
4       Angola  486.0  486.0  486.0  486.0  486.0  486.0  486.0  486.0  486.0   

   ...   2091   2092   2093   2094   2095   2096   2097   2098   2099   2100  
0  ...  12.60  12.40  12.20  12.00  11.80  11.60  11.50  11.30  11.10  11.10  
1  ...   2.32   2.30   2.27   2.24   2.22   2.19   2.16   2.14   2.11   2.11  
2  ...   4.28   4.20   4.14   4.07   4.00   3.94   3.88   3.82   3.76   3.76  
3  ...   0.86   0.84   0.83   0.81   0.80   0.79   0.78   0.77   0.76   0.76  
4  ...  17.70  17.50  17.30  17.10  17.00  16.80  16.60  16.40  16.30  16.30  

[5 rows x 302 columns]

In [31]:
# check data types
df_mortality.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195 entries, 0 to 194
Columns: 302 entries, country to 2100
dtypes: float64(301), object(1)
memory usage: 460.2+ KB


The above output shows that the countries are stored as strings and the mortality rates are stored as floats for all countries. No adjustments needed here. 
***

In [32]:
# check for missing values
sum(df_mortality.isnull().any())

150

This output shows that there are 150 missing values in the data. Since I can see above that there are NaNs for the country Andorra, I am wondering if all the 150 missing values are in that row. 
***

In [33]:
# check missing values for the first 150 data columns of the Andorra row
sum(df_mortality.iloc[3,1:151].isnull())

150

This shows that Andorra is the only country with missing values. As there is no fertility data for Andorra, I will not need this row for my analysis.
***

In [88]:
# load income data
df_income = pd.read_csv('income_per_person.csv')
df_income.head()

country  1800  1801  1802  1803  1804  1805  1806  1807  1808  ...  \
0  Afghanistan   603   603   603   603   603   603   603   603   603  ...   
1      Albania   667   667   667   667   667   668   668   668   668  ...   
2      Algeria   715   716   717   718   719   720   721   722   723  ...   
3      Andorra  1200  1200  1200  1200  1210  1210  1210  1210  1220  ...   
4       Angola   618   620   623   626   628   631   634   637   640  ...   

    2031   2032   2033   2034   2035   2036   2037   2038   2039   2040  
0   2550   2600   2660   2710   2770   2820   2880   2940   3000   3060  
1  19400  19800  20200  20600  21000  21500  21900  22300  22800  23300  
2  14300  14600  14900  15200  15500  15800  16100  16500  16800  17100  
3  73600  75100  76700  78300  79800  81500  83100  84800  86500  88300  
4   6110   6230   6350   6480   6610   6740   6880   7020   7160   7310  

[5 rows x 242 columns]

In [35]:
# check data types
df_income.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 193 entries, 0 to 192
Columns: 242 entries, country to 2040
dtypes: int64(241), object(1)
memory usage: 365.0+ KB


The above output shows that the countries are stored as strings and the income data is stored as ints for all countries. No adjustments needed. 
***

In [36]:
# check for missing values
sum(df_income.isnull().any())

0

The above output shows that there is no missing values in the data.
***

In [89]:
# load education data
df_edu = pd.read_csv('education.csv')
df_edu.head()

country  1970  1971  1972  1973  1974  1975  1976  1977  1978  \
0          Afghanistan   0.1   0.1   0.1   0.1   0.1   0.1   0.1   0.1   0.2   
1              Albania   5.6   5.7   5.9   6.0   6.2   6.3   6.5   6.6   6.8   
2              Algeria   1.4   1.5   1.6   1.7   1.8   1.9   2.1   2.2   2.3   
3               Angola   0.9   1.0   1.0   1.1   1.1   1.2   1.2   1.3   1.4   
4  Antigua and Barbuda   8.1   8.3   8.5   8.7   8.8   9.0   9.2   9.4   9.6   

   ...  2000  2001  2002  2003  2004  2005  2006  2007  2008  2009  
0  ...   0.5   0.5   0.5   0.6   0.6   0.6   0.7   0.7   0.7   0.8  
1  ...   9.8   9.9  10.0  10.1  10.2  10.3  10.4  10.5  10.6  10.7  
2  ...   5.8   5.9   6.1   6.2   6.4   6.5   6.7   6.8   6.9   7.1  
3  ...   3.5   3.6   3.7   3.8   3.9   4.0   4.1   4.3   4.4   4.5  
4  ...  12.7  12.8  12.9  13.0  13.1  13.2  13.3  13.3  13.4  13.5  

[5 rows x 41 columns]

In [38]:
# check data types
df_edu.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 174 entries, 0 to 173
Data columns (total 41 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   country  174 non-null    object 
 1   1970     174 non-null    float64
 2   1971     174 non-null    float64
 3   1972     174 non-null    float64
 4   1973     174 non-null    float64
 5   1974     174 non-null    float64
 6   1975     174 non-null    float64
 7   1976     174 non-null    float64
 8   1977     174 non-null    float64
 9   1978     174 non-null    float64
 10  1979     174 non-null    float64
 11  1980     174 non-null    float64
 12  1981     174 non-null    float64
 13  1982     174 non-null    float64
 14  1983     174 non-null    float64
 15  1984     174 non-null    float64
 16  1985     174 non-null    float64
 17  1986     174 non-null    float64
 18  1987     174 non-null    float64
 19  1988     174 non-null    float64
 20  1989     174 non-null    float64
 21  1990     174 non

The above output shows that the countries are stored as strings and the education data is stored as floats for all countries. Furthermore it turns out that there is only data for 1970-2009. Concerning education, my analysis will be limited to these years. 

I can see in the ouput that there are no missing values.
***

In [90]:
# load urban population data
df_urban = pd.read_csv('urban_population.csv')
df_urban.head()

country   1960    1961    1962    1963    1964   1965   1966   1967  \
0  Afghanistan  0.084  0.0868  0.0898  0.0928  0.0959  0.099  0.102  0.106   
1      Albania  0.307  0.3090  0.3100  0.3110  0.3120  0.312  0.313  0.314   
2      Algeria  0.305  0.3180  0.3320  0.3470  0.3610  0.376  0.388  0.390   
3      Andorra  0.585  0.6100  0.6350  0.6590  0.6820  0.704  0.726  0.746   
4       Angola  0.104  0.1080  0.1120  0.1160  0.1210  0.125  0.130  0.134   

    1968  ...   2010   2011   2012   2013   2014   2015   2016   2017   2018  \
0  0.109  ...  0.237  0.239  0.242  0.244  0.246  0.248  0.250  0.253  0.255   
1  0.314  ...  0.522  0.532  0.543  0.554  0.564  0.574  0.584  0.594  0.603   
2  0.392  ...  0.675  0.682  0.689  0.696  0.702  0.708  0.715  0.721  0.726   
3  0.766  ...  0.888  0.887  0.886  0.885  0.884  0.883  0.882  0.882  0.881   
4  0.139  ...  0.598  0.605  0.613  0.620  0.627  0.634  0.641  0.648  0.655   

    2019  
0  0.258  
1  0.612  
2  0.732  
3  0.880  
4  0.662  

[5 rows x 61 columns]

In [40]:
# check data types
df_urban.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 194 entries, 0 to 193
Data columns (total 61 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   country  194 non-null    object 
 1   1960     192 non-null    float64
 2   1961     192 non-null    float64
 3   1962     192 non-null    float64
 4   1963     192 non-null    float64
 5   1964     192 non-null    float64
 6   1965     192 non-null    float64
 7   1966     192 non-null    float64
 8   1967     192 non-null    float64
 9   1968     192 non-null    float64
 10  1969     192 non-null    float64
 11  1970     192 non-null    float64
 12  1971     192 non-null    float64
 13  1972     192 non-null    float64
 14  1973     192 non-null    float64
 15  1974     192 non-null    float64
 16  1975     192 non-null    float64
 17  1976     192 non-null    float64
 18  1977     192 non-null    float64
 19  1978     192 non-null    float64
 20  1979     192 non-null    float64
 21  1980     192 non

The above output shows that the countries are stored as strings and the urbanity data is stored as floats for all countries. Furthermore it turns out that there is only data for 1960-2019. Concerning urbanity, my analysis will be limited to these years.

I can see that there is data for 194 countries between 1990 and 2011 but only for 192 and 193 countries before and after that time period. Since there is fertility data for only 184 countries this will not be a problem for the analysis.
***

In [91]:
# load geographical data
df_geo = pd.read_csv('country_regions.csv',sep=';')
df_geo.head()

geo         name four_regions       eight_regions  \
0  afg  Afghanistan         asia           asia_west   
1  alb      Albania       europe         europe_east   
2  dza      Algeria       africa        africa_north   
3  and      Andorra       europe         europe_west   
4  ago       Angola       africa  africa_sub_saharan   

                six_regions members_oecd_g77  Latitude Longitude  \
0                south_asia              g77        33        66   
1       europe_central_asia           others        41        20   
2  middle_east_north_africa              g77        28         3   
3       europe_central_asia           others  42,50779   1,52109   
4        sub_saharan_africa              g77     -12,5      18,5   

  UN member since           World bank region  \
0      19.11.1946                  South Asia   
1      14.12.1955       Europe & Central Asia   
2       8.10.1962  Middle East & North Africa   
3       28.7.1993       Europe & Central Asia   
4       1.12.1976          Sub-Saharan Africa   

  World bank, 4 income groups 2017  World bank, 3 income groups 2017  
0                       Low income                               NaN  
1              Upper middle income                               NaN  
2              Upper middle income                               NaN  
3                      High income                               NaN  
4              Lower middle income                               NaN

In [49]:
# check data types in the columns that will be used
df_geo[['name','eight_regions']].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 197 entries, 0 to 196
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   name           197 non-null    object
 1   eight_regions  197 non-null    object
dtypes: object(2)
memory usage: 3.2+ KB


This output shows that both the country names and the region categories are stored as strings and that there are no missing values. No adjustments needed.

### Data Cleaning

There are three data cleaning steps that I will perform.

1. I am interested in the time period between 1800 and 2020. This means I will drop the columns for later years. As already noted above, the education and urbanity data does not cover this whole time period but this just means I need to limit my analysis in those cases. 

2. In the geography data frame I only want to keep the countries and eight_regions columns, so I will drop the other ones.

3. The basic properties of the data displayed above shows that there is fertility data for 184 countries. I want to merge this data frame with the geography data frame and keep only those countries for which I have entries in both data frames.

4. The above basic properties showed that the four indicators have data for more countries than the fertility data. I will thus filter those data frames and keep only the countries that are also contained in the fertility data frame.

In [92]:
# 1. drop years later than 2020 in the respective data frames 
df_fertility.drop(df_fertility.columns[222:],axis=1,inplace=True)
df_mortality.drop(df_mortality.columns[222:],axis=1,inplace=True)
df_income.drop(df_income.columns[222:],axis=1,inplace=True)

In [93]:
# 2. re-write geography data frame, only keeping country and eight_regions
df_geo = df_geo[['name','eight_regions']]
df_geo.rename(columns={'name':'country'},inplace=True)

In [94]:
# 3. merge data
df_fertility = pd.merge(df_geo, df_fertility, on="country", how="inner")

In [148]:
# 4. filter indicator data
countries = df_fertility.country
df_mortality = df_mortality[df_mortality.country.isin(countries)]
df_income = df_income[df_income.country.isin(countries)]
df_edu = df_edu[df_edu.country.isin(countries)]
df_urban = df_urban[df_urban.country.isin(countries)]

<a id='eda'></a>
## Exploratory Data Analysis

In this exploratory data analysis, I will try to answer the questions raised in the introduction.

### How has the global fertility rate changed over time?

### Is the global fertility rate correlated with the four chosen indicators?

### How is the correlation different in various parts of the world?

<a id='conclusions'></a>
## Conclusions
